File: algorithm_server/python/algorithm_worker/cache/metrics.py
from __future__ import annotations

import copy
import os
import threading
import time
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, Callable

from ..collectors.prometheus_collector import PrometheusCollector
from .static_nodes import canonical_hash


@dataclass(frozen=True)
class MetricSnapshot:
    snapshot_id: str
    captured_at: float
    nodes: dict[str, dict[str, float]]


class MetricsCache:
    """Process-local current/previous Prometheus metric snapshots."""

    def __init__(
        self,
        prometheus_url: str = "",
        interval_seconds: float = 30,
        stale_after_seconds: float = 120,
        node_label: str = "node",
        cpu_query: str = "",
        memory_query: str = "",
        request_timeout_seconds: float = 5,
        query: Callable[[str], list[dict[str, Any]]] | None = None,
        clock: Callable[[], float] = time.time,
    ) -> None:
        self.prometheus_url = prometheus_url.rstrip("/")
        self.interval_seconds = max(1.0, interval_seconds)
        self.stale_after_seconds = max(self.interval_seconds, stale_after_seconds)
        self.node_label = node_label
        self.cpu_query = cpu_query
        self.memory_query = memory_query
        collector = PrometheusCollector(
            self.prometheus_url,
            request_timeout_seconds,
        )
        self._query = query or collector.query
        self._clock = clock
        self._lock = threading.RLock()
        self._current: MetricSnapshot | None = None
        self._previous: MetricSnapshot | None = None
        self._last_error = ""
        self._last_query_at = 0.0
        self._stop = threading.Event()
        self._thread: threading.Thread | None = None

    @classmethod
    def from_env(cls) -> "MetricsCache":
        return cls(
            prometheus_url=os.getenv("PROMETHEUS_URL", ""),
            interval_seconds=float(os.getenv("PROMETHEUS_REFRESH_SECONDS", "30")),
            stale_after_seconds=float(os.getenv("PROMETHEUS_STALE_SECONDS", "120")),
            request_timeout_seconds=float(
                os.getenv("PROMETHEUS_REQUEST_TIMEOUT_SECONDS", "5")
            ),
            node_label=os.getenv("PROMETHEUS_NODE_LABEL", "node"),
            cpu_query=os.getenv(
                "PROMETHEUS_CPU_QUERY",
                '1 - avg by (node) (rate(node_cpu_seconds_total{mode="idle"}[5m]))',
            ),
            memory_query=os.getenv(
                "PROMETHEUS_MEMORY_QUERY",
                "1 - (node_memory_MemAvailable_bytes / node_memory_MemTotal_bytes)",
            ),
        )

    @property
    def enabled(self) -> bool:
        return bool(self.prometheus_url and (self.cpu_query or self.memory_query))

    def start(self) -> None:
        if not self.enabled or (self._thread and self._thread.is_alive()):
            return
        self._stop.clear()
        self._thread = threading.Thread(
            target=self._run,
            name="prometheus-metrics-cache",
            daemon=True,
        )
        self._thread.start()

    def stop(self) -> None:
        self._stop.set()
        if self._thread is not None:
            self._thread.join(timeout=min(5.0, self.interval_seconds + 1))
        self._thread = None

    def refresh(self) -> MetricSnapshot:
        if not self.enabled:
            raise RuntimeError("Prometheus metrics cache is disabled")
        values: dict[str, dict[str, float]] = {}
        if self.cpu_query:
            self._merge(values, "cpuUsageRatio", self._query(self.cpu_query))
        if self.memory_query:
            self._merge(values, "memoryUsageRatio", self._query(self.memory_query))
        if not values:
            raise RuntimeError("Prometheus returned no Node metrics")
        captured_at = self._clock()
        canonical_nodes = {name: values[name] for name in sorted(values)}
        snapshot = MetricSnapshot(
            snapshot_id=canonical_hash({"nodes": canonical_nodes}),
            captured_at=captured_at,
            nodes=copy.deepcopy(canonical_nodes),
        )
        with self._lock:
            if self._current and self._current.snapshot_id != snapshot.snapshot_id:
                self._previous = self._current
            self._current = snapshot
            self._last_query_at = captured_at
            self._last_error = ""
        return snapshot

    def for_calculation(self) -> tuple[MetricSnapshot | None, bool, list[str]]:
        with self._lock:
            snapshot = self._current
            error = self._last_error
        if not self.enabled:
            return None, True, ["Prometheus metrics cache is disabled"]
        if snapshot is None:
            return None, True, [error or "Prometheus metric snapshot is not ready"]
        age = max(0.0, self._clock() - snapshot.captured_at)
        if age > self.stale_after_seconds:
            warning = f"Prometheus metric snapshot is stale: ageSeconds={round(age, 1)}"
            if error:
                warning += f"; lastError={error}"
            return snapshot, True, [warning]
        return snapshot, False, []

    def status(self) -> dict[str, Any]:
        snapshot, degraded, warnings = self.for_calculation()
        with self._lock:
            previous = self._previous
            last_error = self._last_error
            last_query_at = self._last_query_at
        return {
            "ready": snapshot is not None,
            "enabled": self.enabled,
            "degraded": degraded,
            "currentSnapshotId": (
                snapshot.snapshot_id if snapshot else "metrics-disabled"
            ),
            "previousSnapshotId": previous.snapshot_id if previous else "",
            "capturedAt": (
                self._timestamp(snapshot.captured_at) if snapshot else ""
            ),
            "lastQueryAt": self._timestamp(last_query_at) if last_query_at else "",
            "nodeCount": len(snapshot.nodes) if snapshot else 0,
            "lastError": last_error or None,
            "warnings": warnings,
        }

    def _run(self) -> None:
        while not self._stop.is_set():
            try:
                self.refresh()
            except Exception as exc:
                with self._lock:
                    self._last_query_at = self._clock()
                    self._last_error = str(exc)
            self._stop.wait(self.interval_seconds)

    def _merge(
        self,
        destination: dict[str, dict[str, float]],
        field: str,
        results: list[dict[str, Any]],
    ) -> None:
        for item in results:
            labels = item.get("metric", {})
            node_name = str(labels.get(self.node_label, ""))
            raw_value = item.get("value", [None, None])
            if not node_name or not isinstance(raw_value, list) or len(raw_value) != 2:
                continue
            try:
                value = max(0.0, min(1.0, float(raw_value[1])))
            except (TypeError, ValueError):
                continue
            destination.setdefault(node_name, {})[field] = value

    @staticmethod
    def _timestamp(value: float) -> str:
        return (
            datetime.fromtimestamp(value, tz=timezone.utc)
            .isoformat()
            .replace("+00:00", "Z")
        )

File: algorithm_server/go/algorithm/metrics.go
// metrics.go 从 Prometheus 拉取 Node CPU、内存和网络指标，并维护带时效性的内存快照。
package algorithm

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"math"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"sync"
	"time"
)

type metricsCache struct {
	// current 用于计算，previous 用于观测版本变化；查询失败不清空成功快照。
	mu               sync.RWMutex
	baseURL          string
	definitions      []metricDefinition
	catalogueVersion string
	bearerToken      string
	nodeLabel        string
	interval         time.Duration
	staleAfter       time.Duration
	client           *http.Client
	current          *metricSnapshot
	previous         *metricSnapshot
	lastError        string
	lastWarnings     []string
	lastQuery        time.Time

	// 仅保留给旧配置和现有测试构造器兼容；正式进程使用 definitions。
	cpuQuery    string
	memoryQuery string
}

func (c *metricsCache) configuredDefinitions() []metricDefinition {
	if len(c.definitions) > 0 {
		return c.definitions
	}
	definitions := make([]metricDefinition, 0, 2)
	if c.cpuQuery != "" {
		definitions = append(definitions, metricDefinition{Name: "cpuUsageRatio", Query: c.cpuQuery, Unit: "ratio", Normalization: "ratio", Required: true})
	}
	if c.memoryQuery != "" {
		definitions = append(definitions, metricDefinition{Name: "memoryUsageRatio", Query: c.memoryQuery, Unit: "ratio", Normalization: "ratio", Required: true})
	}
	return definitions
}

func (c *metricsCache) enabled() bool {
	return c.baseURL != "" && len(c.configuredDefinitions()) > 0
}

func (c *metricsCache) run(ctx context.Context) {
	if !c.enabled() {
		log.Printf("component=algorithm event=prometheus_refresh_skipped reason=disabled")
		return
	}
	c.refresh(ctx)
	ticker := time.NewTicker(c.interval)
	defer ticker.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			c.refresh(ctx)
		}
	}
}

func (c *metricsCache) refresh(ctx context.Context) {
	started := time.Now()
	values := map[string]map[string]float64{}
	units := map[string]string{}
	coverage := map[string]int{}
	warnings := []string{}
	var fatalError error
	for _, definition := range c.configuredDefinitions() {
		units[definition.Name] = definition.Unit
		count, err := c.query(ctx, definition, values)
		coverage[definition.Name] = count
		if err == nil && count > 0 {
			continue
		}
		if err == nil {
			err = fmt.Errorf("returned no Node series")
		}
		message := fmt.Sprintf("Prometheus metric %s: %v", definition.Name, err)
		if definition.Required {
			fatalError = fmt.Errorf("required %s", message)
			break
		}
		warnings = append(warnings, message)
	}

	now := time.Now().UTC()
	c.mu.Lock()
	defer c.mu.Unlock()
	c.lastQuery = now
	c.lastWarnings = append([]string(nil), warnings...)
	// 必需指标失败时保留上一份完整快照；可选网络指标失败只进入降级告警。
	if fatalError != nil {
		c.lastError = fatalError.Error()
		log.Printf("component=algorithm event=prometheus_refresh_completed status=failed metricCount=%d nodeCount=%d warningCount=%d elapsedMs=%.3f error=%q",
			len(coverage), len(values), len(warnings), durationMilliseconds(started), c.lastError)
		return
	}
	if len(values) == 0 {
		c.lastError = "Prometheus returned no Node metrics"
		log.Printf("component=algorithm event=prometheus_refresh_completed status=failed metricCount=%d nodeCount=0 warningCount=%d elapsedMs=%.3f error=%q",
			len(coverage), len(warnings), durationMilliseconds(started), c.lastError)
		return
	}
	id, err := canonicalHash(map[string]any{"catalogueVersion": c.catalogueVersion, "nodes": values})
	if err != nil {
		c.lastError = err.Error()
		log.Printf("component=algorithm event=prometheus_refresh_completed status=failed metricCount=%d nodeCount=%d warningCount=%d elapsedMs=%.3f error=%q",
			len(coverage), len(values), len(warnings), durationMilliseconds(started), c.lastError)
		return
	}
	next := &metricSnapshot{
		SnapshotID: id, CapturedAt: now.Format(time.RFC3339Nano), CapturedUnix: float64(now.UnixNano()) / 1e9,
		CatalogueVersion: c.catalogueVersion, Units: units, Coverage: coverage, Nodes: values,
	}
	if c.current != nil && c.current.SnapshotID != id {
		previous := *c.current
		c.previous = &previous
	}
	c.current = next
	c.lastError = ""
	log.Printf("component=algorithm event=prometheus_refresh_completed status=success metricCount=%d nodeCount=%d warningCount=%d snapshotId=%s elapsedMs=%.3f",
		len(coverage), len(values), len(warnings), shortLogID(id), durationMilliseconds(started))
}

func (c *metricsCache) query(ctx context.Context, definition metricDefinition, destination map[string]map[string]float64) (int, error) {
	endpoint := strings.TrimRight(c.baseURL, "/") + "/api/v1/query?" + url.Values{"query": []string{definition.Query}}.Encode()
	request, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return 0, err
	}
	request.Header.Set("Accept", "application/json")
	if c.bearerToken != "" {
		request.Header.Set("Authorization", "Bearer "+c.bearerToken)
	}
	response, err := c.client.Do(request)
	if err != nil {
		return 0, err
	}
	defer response.Body.Close()
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return 0, fmt.Errorf("HTTP %d", response.StatusCode)
	}
	var payload prometheusResponse
	if err := json.NewDecoder(response.Body).Decode(&payload); err != nil {
		return 0, err
	}
	if payload.Status != "success" || payload.Data.ResultType != "vector" {
		return 0, fmt.Errorf("instant query did not return a vector")
	}
	seen := map[string]struct{}{}
	samples := map[string]float64{}
	for _, item := range payload.Data.Result {
		name := item.Metric[c.nodeLabel]
		if name == "" || len(item.Value) != 2 {
			continue
		}
		var raw string
		if err := json.Unmarshal(item.Value[1], &raw); err != nil {
			continue
		}
		value, err := strconv.ParseFloat(raw, 64)
		if err != nil || math.IsNaN(value) || math.IsInf(value, 0) {
			continue
		}
		if definition.Normalization == "ratio" {
			value = math.Max(0, math.Min(1, value))
		} else {
			value = math.Max(0, value)
		}
		if _, duplicate := seen[name]; duplicate {
			return 0, fmt.Errorf("duplicate Node series %q", name)
		}
		samples[name] = value
		seen[name] = struct{}{}
	}
	for name, value := range samples {
		if destination[name] == nil {
			destination[name] = map[string]float64{}
		}
		destination[name][definition.Name] = value
	}
	return len(seen), nil
}

func (c *metricsCache) resolve() (*metricSnapshot, bool, []string) {
	c.mu.RLock()
	defer c.mu.RUnlock()
	if !c.enabled() {
		return nil, true, []string{"Prometheus metrics cache is disabled"}
	}
	if c.current == nil {
		message := c.lastError
		if message == "" {
			message = "Prometheus metric snapshot is not ready"
		}
		return nil, true, append([]string{message}, c.lastWarnings...)
	}
	warnings := append([]string(nil), c.lastWarnings...)
	// 可选网络指标缺失只降低覆盖率，不把仍有必需指标的快照整体判为不可用。
	degraded := false
	if c.lastError != "" {
		warnings = append(warnings, "last required metric refresh failed: "+c.lastError)
		degraded = true
	}
	age := time.Since(time.Unix(0, int64(c.current.CapturedUnix*1e9)))
	if age > c.staleAfter {
		warning := fmt.Sprintf("Prometheus metric snapshot is stale: ageSeconds=%.1f", age.Seconds())
		if c.lastError != "" {
			warning += "; lastError=" + c.lastError
		}
		warnings = append(warnings, warning)
		degraded = true
	}
	copy := *c.current
	return &copy, degraded, warnings
}

func (c *metricsCache) status() map[string]any {
	snapshot, degraded, warnings := c.resolve()
	c.mu.RLock()
	defer c.mu.RUnlock()
	metricNames := []string{}
	for _, definition := range c.configuredDefinitions() {
		metricNames = append(metricNames, definition.Name)
	}
	currentSnapshotID := "metrics-disabled"
	if c.enabled() {
		currentSnapshotID = "metrics-unavailable"
	}
	result := map[string]any{
		"ready": snapshot != nil, "enabled": c.enabled(), "degraded": degraded,
		"catalogueVersion": c.catalogueVersion, "metricNames": metricNames,
		"currentSnapshotId": currentSnapshotID, "previousSnapshotId": "",
		"capturedAt": "", "lastQueryAt": "", "nodeCount": 0,
		"coverage": map[string]int{}, "lastError": nil, "warnings": warnings,
	}
	if snapshot != nil {
		result["currentSnapshotId"] = snapshot.SnapshotID
		result["capturedAt"] = snapshot.CapturedAt
		result["nodeCount"] = len(snapshot.Nodes)
		result["coverage"] = snapshot.Coverage
	}
	if c.previous != nil {
		result["previousSnapshotId"] = c.previous.SnapshotID
	}
	if !c.lastQuery.IsZero() {
		result["lastQueryAt"] = c.lastQuery.Format(time.RFC3339Nano)
	}
	if c.lastError != "" {
		result["lastError"] = c.lastError
	}
	return result
}
